File: python_libs/quick_sort.py
def partition(arr, lo, hi):
	if lo < hi:
		pivot = hi
		pivot_val = arr[pivot]
		i = lo-1
		j = lo
		for x in range(lo, hi):
			if arr[j] <= pivot_val:
				i +=1
				tmp = arr[i]
				arr[i] = arr[j]
				arr[j] = tmp
			j += 1

		i += 1
		tmp = arr[i]
		arr[i] = pivot_val
		arr[pivot] = tmp
		return i

	elif lo == hi:
		return lo


def quick_sort(arr, start, end):
	if start < end:
		part = partition(arr, start, end)
		quick_sort(arr, start, part-1)
		quick_sort(arr, part+1, end)

if __name__=='__main__':
	n_elem = int(raw_input('Enter number of elements\n'))
	arr = map(int, raw_input('Enter the space-separated array\n').split(' '))
	print 'Sorting ....'
	quick_sort(arr, 0, n_elem-1)
	for x in arr:
		print x,

File: c_libs/quick_sort.c
#include <stdlib.h>

int partition(int* arr, int lo, int hi){
	if(lo<hi){
		int pivot = hi;
		int pivotVal = arr[pivot];
		int i=lo-1, j=lo, tmp, x;
		for(x=lo;x<hi;x++){
			if(arr[j]<=pivotVal){
				i++;
				tmp = arr[i];
				arr[i] = arr[j];
				arr[j] = tmp;
			}
			j++;
		}
		i++;
		tmp = arr[i];
		arr[i] = pivotVal;
		arr[pivot] = tmp;
		return i;
	}
	else if(lo==hi){
		return lo;
	}
}

void quickSort(int* arr, int start, int end){
	if(start<end){
		int part = partition(arr, start, end);
		quickSort(arr, start, part-1);
		quickSort(arr, part+1, end);
	}
}

int main(){
	int n_elem, elem;
	printf("Enter number of elements\n");
	scanf("%d", &n_elem);
	int *arr = (int *)malloc(n_elem*sizeof(int));
	printf("Enter the elements one per line\n");
	for(elem=0; elem<n_elem; elem++){
		scanf("%d", &arr[elem]);
	}
	quickSort(arr, 0, n_elem-1);
	for(elem=0; elem<n_elem; elem++){
		printf("%d", arr[elem]);
	}
	printf("\n");
}